### DISCUSSION TITLE
How to achieve reusable and abstracted font tokens

### DISCUSSION THREAD

**Discussion by arghyab0** [facebook/stylex#335] [Q&A]
Hey folks, I've just started working with StyleX and am having a lot of fun. However, I've run into an issue and would be super grateful for help from the community. I have different font tokens like -

```tsx
displayLarge: {
    fontFamily: fontFamilies.arial, //this is a variable exported from another file
    fontSize: '48px',
    fontWeight: 500,
    fontStyle: fontStyles.normal, //this is a variable exported from another file
    fontFeatureSettings: fontFeatureSettingValues.normal, //this is a variable exported from another file
    lineHeight: '54px',
    letterSpacing: '0.6px',
  },
  displayRegular: {
    fontFamily: fontFamilies.arial,
    fontSize: '36px',
   ...
  },
  ...
```
Now I want to apply these font tokens across different elements. My requirements and ideal setup are -
1. Different elements will have different font tokens
2. Font tokens will be different for an element in desktop vs mobile
3. I want the font token properties to be abstracted and reusable. I should ideally just be able to refer to `displayRegular` (as a group) and it should have all the styling properties; I dont want to individually write the `fontFamily`, `fontSize`, etc
4. I want the least amount of StyleX classes to be applied in markup in an effort to keep it "cleaner" and let the StyleX classes/variables have all the styling info instead

Based on all of these requirements, here's the setup that I came up with -
```tsx
<div {...stylex.props(styles.container, fonts.displayLarge)}>
    <h2>...</h2>
    <p>...</p>
    <SomeComponent />
</div>
```
where `styles.container` and `fonts.displayRegular` are both StyleX classes like -
```tsx
export const styles = stylex.create({
  container: {
    margin: '2rem auto',
    borderWidth: '2px',
    ...
  },
});

```
```tsx
export const fonts = stylex.create({
  displayLarge: {
    fontFamily: fontFamilies.halyardDisplay,
    fontSize: '48px',
    fontWeight: 500,
    fontStyle: fontStyles.normal,
    ...
  },
  displayRegular: {
    fontFamily: fontFamilies.halyardDisplay,
    ...
   }
   ...
});
```
And, I plan to reuse the `fonts` tokens across my codebase. However, even though this meets requirements # 1, # 3 and # 4 in varying degress, # 2 is unachievable without relying on conditional classes.
```tsx
<div {...stylex.props(styles.container, fonts.displayLarge, isMobile && fonts.displayRegular)}>
    <h2>...</h2>
    <p>...</p>
</div>
```
I really don't want to do this because, firstly, I dont want to rely on conditional checks for styling that should have been done through media queries and, secondly, it adds one more StyleX class to the markup. 
Is there a way/pattern to achieve all the 4 requirements that I have? Thanks in advance!

**Comment by nmn** [comment]
Just put the media queries you need within `fonts.displayLarge` itself.

```tsx
const MOBILE = '@media (max-width: 640px)';

export const fonts = stylex.create({
  regular: {
      fontFamily: fontFamilies.arial,
      fontSize: {default: 48, [MOBILE]: 36},
      fontWeight: 500,
      fontStyle: fontStyles.normal,
      fontFeatureSettings: fontFeatureSettingValues.normal,
      lineHeight: {default: 1.125, [MOBILE]: 1.25},
      letterSpacing: {default: '0.6px', [MOBILE]: '1px'},
    },
});
```

You can do this for all your font styles so they adapt based on media query.

If you sometimes need styles that don't change based on media query, define a separate non-responsive style object as well.

**Reply by arghyab0** [reply]
Apologies for not being more specific and providing clearer details initially. I don't have a consistent mapping of properties between desktop and mobile for a specific token (e.g., `fontSize` will always be 48 on desktop and 36 on mobile for the element with the font token `paraLarge`). Nor is there is no consistent mapping of font tokens themselves between desktop and mobile (e.g., if an element has the font token `paraLarge` on desktop, then it must have the font token `paraMedium` on mobile).

My use case is a bit more customised and inconsistent. One element will have two font tokens – one for desktop and one for mobile. The latter can be any token. There may be elements with the token `paraLarge` for desktop and `paraRegular` for mobile. However, there may also be elements with the token `paraLarge` for desktop but `paraExtraSmall` for mobile.

**Reply by nmn** [reply]
StyleX explicitly disallows merging media queries together. The `fontSize` is one value. The `fontWeight` is one value and so on.

The solution to your problem would be to create "classes" for all the possible combinations:

```tsx
export const fonts = stylex.create({
  largeRegular: {
      fontFamily: fontFamilies.arial,
      fontSize: {default: 48, [MOBILE]: 36},
      fontWeight: 500,
      fontStyle: fontStyles.normal,
      fontFeatureSettings: fontFeatureSettingValues.normal,
      lineHeight: {default: 1.125, [MOBILE]: 1.25},
      letterSpacing: {default: '0.6px', [MOBILE]: '1px'},
    },
    largeSmall: {...},
    regular: {...},
    regularSmall: {...},
    regularLarge: {...},
    small: {...},
    smallLarge: {...},
    smallRegular: {...},
    ...
});
```

I understand that this would be a bit verbose. You can use local constants to help a bit, but I'm also working on supporting simple local functions to make this even easier.

---

You also might want to look at using themes instead, which would reduce the amount of verbosity, but I'm not sure exactly how you're using these styles.

---

NIT: Can we not overload the term "token", because a token is usually a reference to a CSS variable.
